modules/memory-management/src/main/java/org/mule/runtime/internal/memory/bytebuffer/WeavePoolBasedByteBufferProvider.java

```diff
@@ -0,0 +1,180 @@
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.internal.memory.bytebuffer;
+
+import static org.mule.runtime.api.profiling.type.RuntimeProfilingEventTypes.MEMORY_BYTE_BUFFER_ALLOCATION;
+import static org.mule.runtime.api.profiling.type.RuntimeProfilingEventTypes.MEMORY_BYTE_BUFFER_DEALLOCATION;
+
+import static java.lang.System.currentTimeMillis;
+
+import org.mule.runtime.api.memory.provider.ByteBufferProvider;
+import org.mule.runtime.api.profiling.ProfilingDataProducer;
+import org.mule.runtime.api.profiling.ProfilingService;
+import org.mule.runtime.api.profiling.type.context.ByteBufferProviderEventContext;
+import org.mule.runtime.internal.memory.bytebuffer.profiling.ContainerProfilingScope;
+import org.mule.runtime.internal.memory.bytebuffer.profiling.DefaultByteBufferProviderEventContext;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A {@link org.mule.runtime.api.memory.provider.ByteBufferProvider} implementation that can be used to retrieve direct or heap
+ * {@link ByteBuffer}'s based on a pool that saves released DirectByteBuffers to be reused.
+ *
+ * Based on DataWeave Implementation.
+ *
+ * @since 4.5.0
+ */
+public class WeavePoolBasedByteBufferProvider implements ByteBufferProvider<ByteBuffer> {
+
+  private final int capacity;
+  private final int maxSize;
+  private final String name;
+
+  private final WeaveByteBufferPool pool;
+  ProfilingDataProducer<ByteBufferProviderEventContext, Object> allocationDataProducer;
+  ProfilingDataProducer<ByteBufferProviderEventContext, Object> deallocationDataProducer;
+
+  protected WeavePoolBasedByteBufferProvider(String name, int capacity, int maxSize, ProfilingService profilingService) {
+    this.name = name;
+    this.capacity = capacity;
+    this.maxSize = maxSize;
+    this.pool = new WeaveByteBufferPool();
+    allocationDataProducer =
+        profilingService.getProfilingDataProducer(MEMORY_BYTE_BUFFER_ALLOCATION, new ContainerProfilingScope());
+    deallocationDataProducer =
+        profilingService.getProfilingDataProducer(MEMORY_BYTE_BUFFER_DEALLOCATION, new ContainerProfilingScope());
+  }
+
+  @Override
+  public ByteBuffer allocate(int size) {
+    allocationDataProducer.triggerProfilingEvent(new DefaultByteBufferProviderEventContext(name, currentTimeMillis(), capacity));
+    ByteBuffer byteBuffer = pool.take();
+    if (byteBuffer == null) {
+      // Takes heap memory
+      return ByteBuffer.allocate(capacity);
+    } else {
+      // Takes direct memory
+      return byteBuffer;
+    }
+  }
+
+  @Override
+  public void release(ByteBuffer buffer) {
+    deallocationDataProducer
+        .triggerProfilingEvent(new DefaultByteBufferProviderEventContext(name, currentTimeMillis(), buffer.limit()));
+    if (buffer.isDirect() && buffer.capacity() == capacity) {
+      // Release direct memory.
+      pool.release(buffer);
+    } else {
+      // Release heap memory.
```

Comment:
Should I create a task about ByteBufferProviderEventContext to handle the type of memory allocated / released? (Direct or heap) So we can differentiate what kind of memory is being used.

Reply:
Yes, create one to add that info for profiling.